pkg/webhook/cvc.go

```diff
@@ -0,0 +1,199 @@
+// provisioning the volume else return nil
+func validateReplicaCount(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	if cvcOldObj.Spec.ReplicaCount != cvcNewObj.Spec.ReplicaCount {
+		return errors.Errorf(
+			"cvc %s replicaCount got modified from %d to %d",
+			cvcOldObj.Name,
+			cvcOldObj.Spec.ReplicaCount,
+			cvcNewObj.Spec.ReplicaCount,
+		)
+	}
+	return nil
+}
+
+// validatePoolListChanges returns error if user modified only the pool names
+func validatePoolListChanges(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	oldDesiredPoolNames := cvc.GetDesiredReplicaPoolNames(cvcOldObj)
+	newDesiredPoolNames := cvc.GetDesiredReplicaPoolNames(cvcNewObj)
+	modifiedPoolNames := util.ListDiff(oldDesiredPoolNames, newDesiredPoolNames)
+	if len(newDesiredPoolNames) >= len(oldDesiredPoolNames) {
+		// If no.of pools on new spec >= no.of pools in old spec(scaleup as well
+		// as migration then there all the pools in old spec must present in new
+		// spec)
+		if len(modifiedPoolNames) > 0 {
+			return errors.Errorf(
+				"volume replica migration directly by modifying pool names %v is not yet supported",
+				modifiedPoolNames,
+			)
+		}
+	} else {
+		// If no.of pools in new spec < no.of pools in old spec(scale down
+		// volume replica case) then there should at most one change in
+		// oldSpec.PoolInfo - newSpec.PoolInfo
+		if len(modifiedPoolNames) > 1 {
+			return errors.Errorf(
+				"Can't perform more than one replica scale down requested scale down count %d",
+				len(modifiedPoolNames),
+			)
+		}
+	}
+	return nil
+}
+
+// validateReplicaScaling returns error if user updated pool list when scaling is
+// already in progress
+func validateReplicaScaling(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	if isScalingInProgress(cvcOldObj) {
+		// if old and new CVC has same count of pools then return true else
+		// return false
+		if len(cvcOldObj.Spec.Policy.ReplicaPoolInfo) != len(cvcNewObj.Spec.Policy.ReplicaPoolInfo) {
+			return errors.Errorf("scaling of CVC %s is already in progress", cvcOldObj.Name)
+		}
+	}
+	return nil
+}
+
+// validateStatusPoolList return error if pools under status doesn't exist under
+// existing spec else return nil
+func validateStatusPoolList(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	replicaPoolNames := []string{}
+	if len(cvcOldObj.Spec.Policy.ReplicaPoolInfo) == 0 &&
+		len(cvcOldObj.Status.PoolInfo) == 0 {
+		// Might be a case where controller updating Bound status along with
+		// spec and status pool names
+		replicaPoolNames = cvc.GetDesiredReplicaPoolNames(cvcNewObj)
+	} else {
+		replicaPoolNames = cvc.GetDesiredReplicaPoolNames(cvcOldObj)
+	}
+	// get pool names which are in status but not under spec
+	invalidStatusPoolNames := util.ListDiff(cvcNewObj.Status.PoolInfo, replicaPoolNames)
+	if len(invalidStatusPoolNames) > 0 {
+		return errors.Errorf(
+			"replica status pool names %v doesn't exist under spec pool list",
```

Comment:
can we return actual validation error i.e. we can modify the existing poolname, only add new pool names ?